src/mlpack/core/cv/purged_k_fold_cv_impl.hpp

```diff
@@ -0,0 +1,376 @@
+        const PredictionsType& ys,
+        const size_t numClasses,
+        const WeightsType& weights,
+        double embargoPercentage,
+        const MatType& intervals) :
+  KFoldCVBase<
+        PurgedKFoldCV<MLAlgorithm, Metric, MatType, PredictionsType, WeightsType>,
+        MLAlgorithm,
+        Metric,
+        MatType,
+        PredictionsType,
+        WeightsType>(k, xs, datasetInfo, ys, numClasses, weights),
+  embargoPercentage(embargoPercentage),
+  intervals(intervals)
+{
+  CheckIntervals();
+}
+
+template<
+    typename MLAlgorithm,
+    typename Metric,
+    typename MatType,
+    typename PredictionsType,
+    typename WeightsType>
+size_t PurgedKFoldCV<
+    MLAlgorithm,
+    Metric,
+    MatType,
+    PredictionsType,
+    WeightsType>::ValidationSubsetFirstCol(size_t i) const
+{
+  assert(i < this->k);
+
+  return (this->k - i - 1) * this->binSize;
+}
+
+template<
+    typename MLAlgorithm,
+    typename Metric,
+    typename MatType,
+    typename PredictionsType,
+    typename WeightsType>
+template<typename ElementType>
+arma::Mat<ElementType> PurgedKFoldCV<
+    MLAlgorithm,
+    Metric,
+    MatType,
+    PredictionsType,
+    WeightsType>::GetTrainingSubset(const arma::Mat<ElementType>& m, size_t i)
+{
+  return m.cols(GetTrainingSubsetCols(i));
+}
+
+template<
+    typename MLAlgorithm,
+    typename Metric,
+    typename MatType,
+    typename PredictionsType,
+    typename WeightsType>
+template<typename ElementType>
+arma::Row<ElementType> PurgedKFoldCV<
+    MLAlgorithm,
+    Metric,
+    MatType,
+    PredictionsType,
+    WeightsType>::GetTrainingSubset(const arma::Row<ElementType>& r, size_t i)
+{
+  return r.cols(GetTrainingSubsetCols(i));
+}
+
+template<
+    typename MLAlgorithm,
+    typename Metric,
+    typename MatType,
+    typename PredictionsType,
+    typename WeightsType>
+template<typename ElementType>
+arma::Mat<ElementType> PurgedKFoldCV<
+    MLAlgorithm,
+    Metric,
+    MatType,
+    PredictionsType,
+    WeightsType>::GetValidationSubset(const arma::Mat<ElementType>& m, size_t i)
+{
+  return m.cols(GetValidationSubsetCols(i));
+}
+
+template<
+    typename MLAlgorithm,
+    typename Metric,
+    typename MatType,
+    typename PredictionsType,
+    typename WeightsType>
+template<typename ElementType>
+arma::Row<ElementType> PurgedKFoldCV<
+    MLAlgorithm,
+    Metric,
+    MatType,
+    PredictionsType,
+    WeightsType>::GetValidationSubset(const arma::Row<ElementType>& r, size_t i)
+{
+  return r.cols(GetValidationSubsetCols(i));
+}
+
+template<
+    typename MLAlgorithm,
+    typename Metric,
+    typename MatType,
+    typename PredictionsType,
+    typename WeightsType>
+void PurgedKFoldCV<
+    MLAlgorithm,
+    Metric,
+    MatType,
+    PredictionsType,
+    WeightsType>::CheckIntervals() const
+{
+  if (intervals.n_rows != 2)
+    throw std::invalid_argument(
+        "PurgedKFoldCV::CheckIntervals(): "
+        "intervals must be a 2 x m matrix!");

```

Comment:
We can only check the number of rows here, but not the number of columns.  I wonder if it would be better to rename this function `CheckState(xs, ys)`, add a version to `KFoldCV` that does nothing, then call `CheckState()` at the start of `TrainAndEvaluate()`.  Otherwise the user could get an out-of-bounds error with seemingly no explanation, if they pass an intervals matrix with a number of columns that does not match `xs`.

Reply:
Implemented the CheckState per your suggestion